internal/middleware/logging.go

```diff
@@ -2,25 +2,56 @@ package middleware
 
 import (
 	"context"
+	"encoding/json"
 	"time"
 
+	"github.com/google/uuid"
+	"github.com/grpc-ecosystem/go-grpc-middleware/logging/zap/ctxzap"
+	"github.com/oklog/ulid/v2"
 	"github.com/openfga/openfga/pkg/logger"
 	serverErrors "github.com/openfga/openfga/pkg/server/errors"
+	"go.opentelemetry.io/otel/trace"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
 )
 
 func NewLoggingInterceptor(logger logger.Logger) grpc.UnaryServerInterceptor {
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
 		start := time.Now()
 
-		resp, err := handler(ctx, req)
+		id, err := uuid.NewRandom()
+		if err != nil {
+			logger.Error("failed to generate uuid", zap.Error(err))
+		}
+		requestID := id.String()
+
+		err = grpc.SetHeader(ctx, metadata.Pairs("X-Request-Id", requestID))
+		if err != nil {
+			logger.Error("failed to set header", zap.Error(err))
+		}
 
 		fields := []zap.Field{
 			zap.String("method", info.FullMethod),
-			zap.Duration("took", time.Since(start)),
+			zap.String("request_id", requestID),
 		}
 
+		spanCtx := trace.SpanContextFromContext(ctx)
+		if spanCtx.HasTraceID() {
+			fields = append(fields, zap.String("trace_id", spanCtx.TraceID().String()))
+		}
+
+		fields = append(fields, ctxzap.TagsToFields(ctx)...)
+
+		jsonReq, err := json.Marshal(req)
+		if err == nil {
+			fields = append(fields, zap.Any("raw_request", json.RawMessage(jsonReq)))
+		}
+
+		resp, err := handler(ctx, req)
+
+		fields = append(fields, zap.Duration("took", time.Since(start)))
```

Comment:
Can we please rename this to `request_duration`? `took` is just too generic.

Reply:
Done!

Reply:
This will break https://github.com/auth0/dwh-dbt-snowflake/blob/master/models/marts/fga/fga_store_logs.sql#L19

Reply:
Oh... good catch!

Reply:
@jon-whit Shall we just leave it as took?

Reply:
Let's inject `request_duration` and `took` (same values, but different names). We can refactor references to `took` so they reference `request_duration` instead, and then we can drop `took`.

Reply:
Added back for now. Thanks!